File: app/src/main/java/cm/aptoide/pt/app/view/donations/DonationsService.java
package cm.aptoide.pt.app.view.donations;

import cm.aptoide.pt.app.view.donations.data.GetDonations;
import java.util.ArrayList;
import java.util.List;
import retrofit2.http.GET;
import retrofit2.http.Query;
import rx.Observable;
import rx.Scheduler;
import rx.Single;

/**
 * Created by franciscocalado on 11/10/2018.
 */

public class DonationsService {

  private ServiceV8 service;
  private Scheduler ioScheduler;

  public DonationsService(ServiceV8 service, Scheduler ioScheduler) {
    this.service = service;
    this.ioScheduler = ioScheduler;
  }

  public Single<List<Donation>> getDonations(String packageName) {
    return service.getDonations(packageName, 5)
        .subscribeOn(ioScheduler)
        .map(this::mapToDonationsList)
        .toSingle();
  }

  private List<Donation> mapToDonationsList(GetDonations donationsResponse) {
    List<Donation> result = new ArrayList<>();

    for (GetDonations.Donor donor : donationsResponse.getItems())
      result.add(
          new Donation(donor.getDomain(), donor.getOwner(), Float.parseFloat(donor.getAppc())));
    return result;
  }

  public interface ServiceV8 {
    @GET("broker/8.20181010/leaderboard/donations") Observable<GetDonations> getDonations(
        @Query("domain") String packageName, @Query("limit") int limit);
  }
}

File: appcoins/src/main/java/cm/aptoide/pt/donations/DonationsService.kt
package cm.aptoide.pt.donations

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import retrofit2.Response
import retrofit2.http.GET
import retrofit2.http.Query
import java.util.*

class DonationsService(private val serviceV8: ServiceV8) {

  suspend fun getDonations(packageName: String): List<Donation> {
    return withContext(Dispatchers.IO) {
      val donationsResponse = serviceV8.getDonations(packageName, 5)
      val donationsResponseBody = donationsResponse.body()
      if (donationsResponse.isSuccessful && donationsResponseBody != null) {
        return@withContext mapToDonationsList(donationsResponseBody)
      } else {
        return@withContext arrayListOf<Donation>()
      }
    }
  }


  private fun mapToDonationsList(
      donationsResponse: GetDonations): List<Donation> {
    val result: MutableList<Donation> = ArrayList<Donation>()
    for (donor in donationsResponse.items) result.add(
        Donation(donor.domain, donor.owner, donor.appc.toFloat()))
    return result
  }

  interface ServiceV8 {
    @GET("broker/8.20181010/leaderboard/donations")
    suspend fun getDonations(
        @Query("domain") packageName: String,
        @Query("limit") limit: Int): Response<GetDonations>
  }
}